dbms/src/Interpreters/Context.cpp

```diff
@@ -1348,15 +1348,21 @@ std::shared_ptr<Cluster> Context::tryGetCluster(const std::string & cluster_name
 }
 
 
+void Context::reloadClusterConfig()
+{
+    std::lock_guard<std::mutex> lock(shared->clusters_mutex);
+    auto & config = shared->clusters_config ? *shared->clusters_config : getConfigRef();
+    shared->clusters = std::make_unique<Clusters>(config, settings);
```

Comment:
It's possible to update before locking a mutex, and do only assignment under mutex, like `MultiVersion`.
... And you can directly use `MultiVersion`.

Reply:
Not so that. Initially, you should read `clusters_config` which is protected by the same mutex. So the code is going to look like:
```c++
ConfigurationPtr current_clusters_config;
{
	std::lock_guard<std::mutex> lock(shared->clusters_mutex);
	current_clusters_config = shared->clusters_config;
}

auto & config = current_clusters_config ? *current_clusters_config : getConfigRef();
auto new_clusters = std::make_unique<Clusters>(config, settings);

{
	std::lock_guard<std::mutex> lock(shared->clusters_mutex);
	shared->clusters = new_clusters;
}
```